--- crossdev-20111118/crossdev.orig
+++ crossdev-20111118/crossdev
@@ -38,7 +38,7 @@
     ${GOOD}--k, --kernel${NORMAL} ver        Specify version of kernel headers to use
     ${GOOD}--l, --libc${NORMAL} ver          Specify version of libc to use
     ${GOOD}-A, --abis${NORMAL} abis          Specify ABIs to build, first one is the default
-    ${GOOD}--[bgkl]env${NORMAL} env          Specify env settings for binutils/gcc/kernel/libc
+    ${GOOD}--[bdgikl]env${NORMAL} env        Specify env settings for binutils/gcc/gdb/insight/kernel/libc
     ${GOOD}-f, --force${NORMAL}              I don't need a seat belt!
     ${GOOD}-S, --stable${NORMAL}             Use latest stable versions as default
     ${GOOD}-C, --clean${NORMAL} target       Uninstall specified target
@@ -52,7 +52,9 @@
     ${GOOD}-oO, --ov-output${NORMAL} path    Overlay to write crossdev package links
                                 [default: last from --overlays list]
     ${GOOD}-ob, --ov-binutils${NORMAL} path  Overlay for binutils ebuilds [default: search]
+    ${GOOD}-od, --ov-gdb${NORMAL} path       Overlay for gdb ebuilds [default: search]
     ${GOOD}-og, --ov-gcc${NORMAL} path       Overlay for gcc ebuilds [default: search]
+    ${GOOD}-oi, --ov-insight${NORMAL} path   Overlay for insight ebuilds [default: search]
     ${GOOD}-ok, --ov-kernel${NORMAL} path    Overlay for kernel ebuilds [default: search]
     ${GOOD}-ol, --ov-libc${NORMAL} path      Overlay for C library ebuilds [default: search]
 
@@ -190,7 +192,14 @@
 		      LCAT="dev-util"; LPKG="mingw-runtime";;
 
 		msp430)
-		      STAGE_DEFAULT=${STAGE_BINUTILS};;
+		      KCAT="dev-embedded"; KPKG="msp430mcu";
+		      BPKG="msp430-binutils";
+		      GPKG="msp430-gcc";
+		      LCAT="dev-embedded"; LPKG="msp430-libc";
+		      DPKG="msp430-gdb";
+		      IPKG="msp430-insight";
+		      GUSE="-fortran -openmp";
+		      WITH_DEF_HEADERS="no";;
 
 		nios2*)
 		      BPKG="binutils-nios2";
@@ -446,6 +455,8 @@
 GCAT="sys-devel"  ; GPKG="gcc"           ; GVER="" GUSE="" GENV="" GOVL=""
 KCAT="sys-kernel" ; KPKG="linux-headers" ; KVER="" KUSE="" KENV="" KOVL=""
 LCAT="sys-libs"   ; LPKG="[none]"        ; LVER="" LUSE="" LENV="" LOVL=""
+DCAT="sys-devel"  ; DPKG="gdb"           ; DVER="" DUSE="" DENV="" DOVL=""
+ICAT="dev-util"   ; DPKG="insight"       ; IVER="" IUSE="" IENV="" IOVL=""
 DEFAULT_VER="[latest]"
 SEARCH_OVERLAYS=""
 CROSSDEV_OVERLAY=""
@@ -469,9 +480,15 @@
 	--b|--binutils)    shift; BVER=$1;;
 	--benv)            shift; BENV=$1;;
 	-ob|--ov-binutils) shift; BOVL=$1;;
+	--d|--gdb)         shift; DVER=$1;;
+	--denv)            shift; DENV=$1;;
+	-od|--ov-gdb)      shift; DOVL=$1;;
 	--g|--gcc)         shift; GVER=$1;;
 	--genv)            shift; GENV=$1;;
 	-og|--ov-gcc)      shift; GOVL=$1;;
+	--i|--insight)     shift; IVER=$1;;
+	--ienv)            shift; IENV=$1;;
+	-oi|--ov-insight)  shift; IOVL=$1;;
 	--k|--kernel)      shift; KVER=$1;;
 	--kenv)            shift; KENV=$1;;
 	-ok|--ov-kernel)   shift; KOVL=$1;;
@@ -811,8 +828,8 @@
 set_portage ${GCAT} ${GPKG} ${GVER} "${GENV}" "${GOVL}"
 set_portage ${KCAT} ${KPKG} ${KVER} "${KENV}" "${KOVL}"
 set_portage ${LCAT} ${LPKG} ${LVER} "${LENV}" "${LOVL}"
-set_portage sys-devel gdb
-set_portage dev-util insight
+set_portage ${DCAT} ${DPKG} ${DVER} "${DENV}" "${DOVL}"
+set_portage ${ICAT} ${IPKG} ${IVER} "${IENV}" "${IOVL}"
 set_metadata
 
 # filter out revdep rebuild stuff #182601
@@ -924,7 +941,7 @@
 
 EOPTS="${EOPTS_UP} --newuse"
 ex_gcc && USE="${GUSE} ${USE}" doemerge ${GPKG} ${GPKG}-extra
-ex_gdb && doemerge gdb
-ex_insight && doemerge insight
+ex_gdb && USE="${DUSE} ${USE}" doemerge ${DPKG}
+ex_insight && USE="${IUSE} ${USE}" doemerge ${IPKG}
 
 exit 0
