--- crossdev-20100108/crossdev.orig	2010-04-14 13:20:30.631080303 -0700
+++ crossdev-20100108/crossdev	2010-04-16 14:02:46.118085525 -0700
@@ -166,7 +166,14 @@
 		      LCAT="dev-util"; LPKG="mingw-runtime";;
 
 		msp430)
-		      STAGE_DEFAULT=${STAGE_BINUTILS};;
+		      KPKG="[none]";
+		      BCAT="sys-devel"; BPKG="msp430-binutils";
+		      GCAT="sys-devel"; GPKG="msp430-gcc";
+		      LCAT="dev-embedded"; LPKG="msp430-libc";
+		      DCAT="sys-devel"; DPKG="msp430-gdb";
+		      ICAT="dev-util"; IPKG="msp430-insight";
+		      GUSE="-fortran -openmp" # doesn't work
+		      WITH_DEF_HEADERS="no";;
 
 		nios2*)
 		      BPKG="binutils-nios2";
@@ -359,6 +366,7 @@
 GCAT="sys-devel"  ; GPKG="gcc"           ; GVER="" ; GUSE=""
 KCAT="sys-kernel" ; KPKG="linux-headers" ; KVER="" ; KUSE=""
 LCAT="sys-libs"   ; LPKG=""              ; LVER="" ; LUSE=""
+DCAT="sys-devel"  ; DPKG="gdb"           ; DVER="" ; DUSE=""
 DEFAULT_VER="[latest]"
 GUSE_DISABLE="-boundschecking -d -fortran -gtk -gcj -libffi -mudflap nocxx -objc -objc++ -objc-gc -openmp"
 GUSE_DISABLE_STAGE_2=${GUSE_DISABLE/-fortran}
@@ -610,8 +618,8 @@
 set_portage ${GCAT} ${GPKG} ${GVER}
 set_portage ${KCAT} ${KPKG} ${KVER}
 set_portage ${LCAT} ${LPKG} ${LVER}
-set_portage sys-devel gdb
-set_portage dev-util insight
+set_portage ${DCAT} ${DPKG} ${DVER}
+set_portage ${ICAT} ${IPKG} ${IVER}
 
 #######################################
 ### Create links for helper scripts ###
@@ -717,7 +725,7 @@
 
 EOPTS="${EOPTS_UP} --newuse"
 ex_gcc && USE="${GUSE} ${USE}" doemerge ${GPKG} ${GPKG}-extra
-ex_gdb && doemerge gdb
-ex_insight && doemerge insight
+ex_gdb && USE="${DUSE} ${USE}" doemerge ${DPKG}
+ex_insight && doemerge ${IPKG}
 
 exit 0
