--- crossdev-20101011/crossdev.orig	2010-11-05 15:22:17.938837729 -0700
+++ crossdev-20101011/crossdev	2010-11-05 15:21:57.503657342 -0700
@@ -151,7 +151,11 @@
 	# Now account for all the "special" architectures out there
 	case ${CTARGET} in
 		# this is a linux+ target, not microcontroller (below)
-		avr32*) :;;
+		avr32*)
+		      KPKG="avr32-linux-headers";
+		      BPKG="avr32-binutils";
+		      GPKG="avr32-gcc";
+		      DPKG="avr32-gdb";;
 
 		avr*) KPKG="[none]";
 		      LCAT="dev-embedded"; LPKG="avr-libc";
