--- msp430-libc-ti_20100815/src/Makefile.orig	2010-08-15 10:03:53.000000000 -0700
+++ msp430-libc-ti_20100815/src/Makefile	2010-09-09 15:02:58.230239742 -0700
@@ -25,7 +25,7 @@
 override AR = msp430-ar
 RM = rm
 MD = mkdir -p $@
-INSTALL = install -c -m644 $^/
+INSTALL = install -c -m644 $^
 
 #HDR_VENDOR := legacy
 HDR_VENDOR := ti
@@ -91,8 +91,9 @@
 clean-crt:
 	${RM} -f $(CRT_OBJS)
 
-install-crt: $(CRT_OBJS) ${libdir}
-	${INSTALL}
+install-crt:
+	mkdir -p ${libdir}
+	${INSTALL} $(CRT_OBJS) ${libdir}
 
 makedir-crt:
 	mkdir -p $(BUILDDIR)
@@ -128,24 +129,29 @@
 
 install-headers-legacy: ins-legacy-msp430
 
-ins-hr: $(wildcard $(SRC_ROOT)/../include/*.h) $(wildcard $(SRC_ROOT)/../include/$(HDR_VENDOR)/*.h) ${includedir}
-	${INSTALL}
-
-ins-hs: $(wildcard $(SRC_ROOT)/../include/sys/*.h) ${includedir}/sys
-	${INSTALL}
-
-ins-hb: $(wildcard $(SRC_ROOT)/../include/bits/*.h) ${includedir}/bits
-	${INSTALL}
-
-ins-legacy-msp430: $(wildcard $(SRC_ROOT)/../include/legacy/msp430/*.h) ${includedir}/msp430
-	${INSTALL}
-
-fix-limits: $(wildcard $(SRC_ROOT)/../include/limits.h) ${includedir}/../../lib/gcc-lib/msp430/3.0/include
-	${INSTALL}	
+ins-hr:
+	mkdir -p ${includedir}
+	${INSTALL} $(wildcard $(SRC_ROOT)/../include/*.h) $(wildcard $(SRC_ROOT)/../include/$(HDR_VENDOR)/*.h) ${includedir}
+
+ins-hs:
+	mkdir -p ${includedir}/sys
+	${INSTALL} $(wildcard $(SRC_ROOT)/../include/sys/*.h) ${includedir}/sys
+
+ins-hb:
+	mkdir -p ${includedir}/bits
+	${INSTALL} $(wildcard $(SRC_ROOT)/../include/bits/*.h) ${includedir}/bits
+
+ins-legacy-msp430:
+	mkdir -p ${includedir}/msp430
+	${INSTALL} $(wildcard $(SRC_ROOT)/../include/legacy/msp430/*.h) ${includedir}/msp430
+
+fix-limits:
+	mkdir -p ${includedir}/../../lib/gcc-lib/msp430/3.0/include
+	${INSTALL} $(wildcard $(SRC_ROOT)/../include/limits.h) ${includedir}/../../lib/gcc-lib/msp430/3.0/include
 
 #--------------- directories ---------------
-$(includedir) $(includedir)/bits $(includedir)/sys $(includedir)/msp430:
-	$(MD)
+#$(includedir) $(includedir)/bits $(includedir)/sys $(includedir)/msp430:
+#	$(MD)
 #------------------------------------------------------------------------
 #------------------------------------------------------------------------
 #------------------------------------------------------------------------
@@ -196,8 +202,8 @@
 clean-libc:
 	${RM} -f $(LIBC) $(LIBC_OBJS)
 
-install-libc: $(LIBC) $(libdir)/$(MULTILIB_DIR)
-	${INSTALL}
+install-libc:
+	${INSTALL} $(LIBC) $(libdir)/$(MULTILIB_DIR)
 #------------------------------------------------------------------------
 #-------------- Math & FP libs ---------------
 
@@ -229,8 +235,8 @@
 $(LIBFP): $(LIBFP_OBJS)
 	${AR} rc $@ $?
 
-install-libm: $(LIBM) $(LIBFP) $(libdir)/$(MULTILIB_DIR)
-	${INSTALL}
+install-libm: 
+	${INSTALL} $(LIBM) $(LIBFP) $(libdir)/$(MULTILIB_DIR)
 
 clean-libm:
 	${RM} -f $(LIBM) $(LIBFP) $(LIBM_OBJS) $(LIBFP_OBJS)
