--- uClibc-0.9.31/Makefile.in.orig	2010-11-06 23:30:38.404036862 -0700
+++ uClibc-0.9.31/Makefile.in	2010-11-06 23:40:18.456455075 -0700
@@ -289,6 +289,8 @@
 ifneq ($(findstring install,$(MAKECMDGOALS)),)
 $(PREFIX)$(RUNTIME_PREFIX)$(MULTILIB_DIR) $(addprefix $(PREFIX)$(DEVEL_PREFIX),include lib):
 	$(do_mkdir)
+$(PREFIX)$(DEVEL_PREFIX)$(MULTILIB_DIR):
+	$(do_mkdir)
 endif
 
 install_headers: headers $(top_builddir)extra/scripts/unifdef | $(PREFIX)$(DEVEL_PREFIX)include
