--- msp430-libc-20100624/src/Makefile.orig	2010-06-25 09:58:58.000000000 +0800
+++ msp430-libc-20100624/src/Makefile	2010-06-25 13:58:45.000000000 +0800
@@ -25,7 +25,7 @@
 override AR = msp430-ar
 RM = rm
 MD = mkdir -p $@
-INSTALL = install -c -m644 $^/
+INSTALL = install -c -m644 $^
 
 COMMON := $(LIB_OPTIONS)
 # make sure we can find our header files (not the installed ones)
@@ -88,8 +88,9 @@
 clean-crt:
 	${RM} -f $(CRT_OBJS)
 
-install-crt: $(CRT_OBJS) ${libdir}
-	${INSTALL}
+install-crt:
+	mkdir -p ${libdir}
+	${INSTALL} $(CRT_OBJS) ${libdir}
 
 makedir-crt:
 	mkdir -p $(BUILDDIR)
@@ -121,24 +122,29 @@
 
 install-headers: ins-hr ins-hs ins-hb ins-msp430
 
-ins-hr: $(wildcard $(SRC_ROOT)/../include/*.h) ${includedir}
-	${INSTALL}
+ins-hr:
+	mkdir -p ${includedir}
+	${INSTALL} $(wildcard $(SRC_ROOT)/../include/*.h) ${includedir}
+
+ins-hs:
+	mkdir -p ${includedir}/sys
+	${INSTALL} $(wildcard $(SRC_ROOT)/../include/sys/*.h) ${includedir}/sys
+
+ins-hb:
+	${INSTALL} $(wildcard $(SRC_ROOT)/../include/bits/*.h) ${includedir}/bits
+
+ins-msp430:
+	mkdir -p ${includedir}/msp430
+	${INSTALL} $(wildcard $(SRC_ROOT)/../include/msp430/*.h) ${includedir}/msp430
+
+fix-limits:
+	mkdir -p ${includedir}/../../lib/gcc-lib/msp430/3.0/include
+	${INSTALL} $(wildcard $(SRC_ROOT)/../include/limits.h) ${includedir}/../../lib/gcc-lib/msp430/3.0/include
 
-ins-hs: $(wildcard $(SRC_ROOT)/../include/sys/*.h) ${includedir}/sys
-	${INSTALL}
-
-ins-hb: $(wildcard $(SRC_ROOT)/../include/bits/*.h) ${includedir}/bits
-	${INSTALL}
-
-ins-msp430: $(wildcard $(SRC_ROOT)/../include/msp430/*.h) ${includedir}/msp430
-	${INSTALL}
-
-fix-limits: $(wildcard $(SRC_ROOT)/../include/limits.h) ${includedir}/../../lib/gcc-lib/msp430/3.0/include
-	${INSTALL}	
 
 #--------------- directories ---------------
-$(includedir) $(includedir)/bits $(includedir)/sys $(includedir)/msp430:
-	$(MD)
+#$(includedir) $(includedir)/bits $(includedir)/sys $(includedir)/msp430:
+#	$(MD)
 #------------------------------------------------------------------------
 #------------------------------------------------------------------------
 #------------------------------------------------------------------------
@@ -189,8 +195,8 @@
 clean-libc:
 	${RM} -f $(LIBC) $(LIBC_OBJS)
 
-install-libc: $(LIBC) $(libdir)/$(MULTILIB_DIR)
-	${INSTALL}
+install-libc:
+	${INSTALL} $(LIBC) $(libdir)/$(MULTILIB_DIR)
 #------------------------------------------------------------------------
 #-------------- Math & FP libs ---------------
 
@@ -222,8 +228,8 @@
 $(LIBFP): $(LIBFP_OBJS)
 	${AR} rc $@ $?
 
-install-libm: $(LIBM) $(LIBFP) $(libdir)/$(MULTILIB_DIR)
-	${INSTALL}
+install-libm: 
+	${INSTALL} $(LIBM) $(LIBFP) $(libdir)/$(MULTILIB_DIR)
 
 clean-libm:
 	${RM} -f $(LIBM) $(LIBFP) $(LIBM_OBJS) $(LIBFP_OBJS)
